README.md
[![Kwalitee status](https://cpants.cpanauthors.org/dist/OPM-Repository.png)](https://cpants.cpanauthors.org/dist/OPM-Repository)
[![GitHub issues](https://img.shields.io/github/issues/perlservices/OPM-Repository.svg)](https://github.com/perlservices/OPM-Repository/issues)
[![CPAN Cover Status](https://cpancoverbadge.perl-services.de/OPM-Repository-1.0.0)](https://cpancoverbadge.perl-services.de/OPM-Repository-1.0.0)
[![Cpan license](https://img.shields.io/cpan/l/OPM-Repository.svg)](https://metacpan.org/release/OPM-Repository)

# NAME

OPM::Repository - parse OPM repositories' framework.xml files to search for add ons

# VERSION

version 1.0.0

# SYNOPSIS

```perl
use OPM::Repository;

my $repo = OPM::Repository->new(
    sources => [qw!
        https://opar.perl-services.de/framework.xml
        https://download.znuny.org/releases/packages/framework.xml
        https://download.znuny.org/releases/itsm/packages6/framework.xml
    !],
);

my ($url) = $repo->find(
  name      => 'ITSMCore',
  framework => '3.3',
);

print $url;
```

## BUILDARGS

# ATTRIBUTES

- sources

# METHODS

## new

`new` has only one mandatory parameter: _sources_. This has to be 
an array reference of URLs for repositories' framework.xml files.

```perl
my $repo = OPM::Repository->new(
    sources => [qw!
        http://opar.perl-services.de/framework.xml
        http://ftp.framework.org/pub/framework/packages/framework.xml
        http://ftp.framework.org/pub/framework/itsm/packages33/framework.xml
    !],
);
```

## find

Search for an add on for a given OPM version in those repositories. It
returns a list of urls if the add on was found, `undef` otherwise.

```perl
my @urls = $repo->find(
  name      => 'ITSMCore',
  framework => '3.3',
);
```

Find a specific version

```perl
my @urls = $repo->find(
  name      => 'ITSMCore',
  framework => '3.3',
  version   => '1.4.8',
);
```

## list

List all addons found in the repositories

```perl
my @addons = $repo->list;
say $_ for @addons;
```

You can also define the OPM version

```perl
my @addons = $repo->list( framework => '5.0.x' );
say $_ for @addons;
```

Both snippets print a simple list of addon names. If you want to
to create a list with more information, you can use

```perl
my @addons = $repo->list(
    framework => '5.0.x',
    details   => 1,
);
say sprintf "%s (%s) on %s\n", $_->{name}, $_->{version}, $_->{url} for @addons;
```



# Development

The distribution is contained in a Git repository, so simply clone the
repository

```
$ git clone git://github.com/perlservices/OPM-Repository.git
```

and change into the newly-created directory.

```
$ cd OPM-Repository
```

The project uses [`Dist::Zilla`](https://metacpan.org/pod/Dist::Zilla) to
build the distribution, hence this will need to be installed before
continuing:

```
$ cpanm Dist::Zilla
```

To install the required prequisite packages, run the following set of
commands:

```
$ dzil authordeps --missing | cpanm
$ dzil listdeps --author --missing | cpanm
```

The distribution can be tested like so:

```
$ dzil test
```

To run the full set of tests (including author and release-process tests),
add the `--author` and `--release` options:

```
$ dzil test --author --release
```

# AUTHOR

Renee Baecker <reneeb@cpan.org>

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2017 by Renee Baecker.

This is free software, licensed under:

```
The Artistic License 2.0 (GPL Compatible)
```


Changes
Change history for OPM::Repository
==================================

0.09 2017-11-16

  + fix tests

0.08 2017-11-16

  [IMPROVEMENTS]
  + add "details" option to list() to get a list of available addons with more details (name, version, url)

  [DOCUMENTATION]
  + add documentation for list() method

0.07 2017-11-05

  + fix dist (updated dist.ini)

0.06 2017-11-05

  + Allow file:// protocol for repositories

0.05 2014-01-21

  + fix test failures on Win32

0.04 2014-01-20

  + better handling for invalid URLs and invalid XML
  + add tests
  + fix test failures on Win32 (Regexp::Common doesn't recognize file://C:/dir/test.xml as valid file-URI)

0.03 2014-01-03

  + fix min required version of List::Util
  + add feature: list names of all packages
  + added metainfo to dist.ini

0.02 2014-01-02

  + add feature: search for a specific version of a module
  + use PodWeaver plugin for dzil

0.01 2014-01-01

  + initial release


lib/OPM/Repository/Source.pm
package OPM::Repository::Source;

use v5.10;

# ABSTRACT: Parser for a single {otrs|otobo}.xml file

use strict;
use warnings;

use Moo;
use HTTP::Tiny;
use HTTP::Tiny::FileProtocol;
use XML::LibXML;
use Regexp::Common qw(URI);

our $VERSION = 0.08;

our $ALLOWED_SCHEME = [ 'HTTP', 'file' ];

has url      => ( is => 'ro', required => 1, isa     => \&_check_uri );
has content  => ( is => 'ro', lazy     => 1, builder => \&_get_content );
has tree     => ( is => 'ro', lazy     => 1, builder => \&_build_tree );
has error    => ( is => 'rwp' );
has packages => ( is => 'rwp', default => sub { {} }, isa => sub { die "No hashref" unless ref $_[0] eq 'HASH' } );
has parsed   => ( is => 'rwp', predicate => 1 );
has product  => ( is => 'ro', default => sub { 'otrs' } );

sub find {
    my ($self, %params) = @_;

    return if !exists $params{name};
    return if !exists $params{framework};

    my $package   = $params{name};
    my $framework = $params{framework};

    if ( !defined $package || !defined $framework ) {
        return;
    }

    if ( !$self->has_parsed ) {
        $self->_parse( %params );
    }

    my %packages = %{ $self->packages };

    return if !$packages{$package};
    return if !$packages{$package}->{$framework};

    my $wanted = $params{version} || $packages{$package}->{$framework}->{latest};
    return $packages{$package}->{$framework}->{versions}->{$wanted};
}

sub list {
    my ($self, %params) = @_;

    if ( !$self->has_parsed ) {
        $self->_parse( %params );
    }

    my %packages  = %{ $self->packages };
    my $framework = $params{framework};

    my @package_names = sort keys %packages;

    if ( $framework ) {
        @package_names = grep{ $packages{$_}->{$framework} }@package_names;
    }

    if ( $params{details} ) {
        my @package_list;

        NAME:
        for my $name ( @package_names ) {
            my @all_framework_versions = $framework ? $framework : keys %{ $packages{$name} || {} };

            OPM_VERSION:
            for my $framework_version ( @all_framework_versions ) {

                VERSION:
                for my $version ( keys %{ $packages{$name}->{$framework_version}->{versions} || {} } ) {
                    push @package_list, {
                        name    => $name,
                        version => $version,
                        url     => $packages{$name}->{$framework_version}->{versions}->{$version},
                    }
                }
            }
        }

        return sort { $a->{name} cmp $b->{name} || $a->{version} cmp $b->{version} } @package_list;
    }

    return @package_names;
}

sub _check_uri {
    my @allowed_schemes = ref $ALLOWED_SCHEME ? @{ $ALLOWED_SCHEME } : $ALLOWED_SCHEME;

    my $matches;

    SCHEME:
    for my $scheme ( @allowed_schemes ) {
        my $regex = ( lc $scheme eq 'http' ) ?
            $RE{URI}{HTTP}{-scheme => qr/https?/} :
            $RE{URI}{$scheme};

        if ( $_[0] =~ m{\A$regex\z} ) {
            $matches++;
            last SCHEME;
        }
    }

    die "No valid URI" unless $matches;
    return 1;
}

sub _parse {
    my ($self, %params) = @_;

    return if !$self->tree;

    my %packages = %{ $self->packages };

    my @repo_packages = $self->tree->findnodes( 'Package' );
    my $base_url      = $self->url;
    $base_url         =~ s{\w+\.xml\z}{};

    REPO_PACKAGE:
    for my $repo_package ( @repo_packages ) {
        my $name       = $repo_package->findvalue( 'Name' );
        my @frameworks = $repo_package->findnodes( 'Framework' );
        my $file       = $repo_package->findvalue( 'File' );

        my $version    = $repo_package->findvalue( 'Version' );

        FRAMEWORK:
        for my $framework ( @frameworks ) {
            my $framework_version  = $framework->textContent;
            my $short_version = join '.', (split /\./, $framework_version, 3)[0..1];
            my $saved_version = $packages{$name}->{$short_version}->{latest};

            my $minimum = $framework->findvalue('@Minimum');
            my $maximum = $framework->findvalue('@Maximum');

            if ( !$saved_version ) {
                $packages{$name}->{$short_version} = {
                    latest       => $version,
                    min_versions => {
                    },
                    max_versions => {
                    },
                    versions     => {
                      $version => sprintf "%s%s", $base_url, $file,
                    },
                };
            }
            elsif ( $self->_version_is_newer( $version, $saved_version ) ) {
                $packages{$name}->{$short_version}->{latest} = $version;
                $packages{$name}->{$short_version}->{versions}->{$version} =
                    sprintf "%s%s", $base_url, $file;
            }
            else {
                $packages{$name}->{$short_version}->{versions}->{$version} =
                    sprintf "%s%s", $base_url, $file;
            }
        }
    }

    $self->_set_parsed( 1 );
    $self->_set_packages( \%packages );

    return 1;
}

sub _version_is_newer {
    my ($self, $new, $old) = @_;

    my @new_levels = split /\./, $new;
    my @old_levels = split /\./, $old;

    for my $i ( 0 .. ( $#new_levels > $#old_levels ? @new_levels : @old_levels ) ) {
        if ( !$old_levels[$i] || $new_levels[$i] > $old_levels[$i] ) {
            return 1;
        }
        elsif ( $new_levels[$i] < $old_levels[$i] ) {
            return 0;
        }
    }

    return 1;
}

sub _get_content {
    my $self = shift;
    my $res  = HTTP::Tiny->new->get( $self->url );

    $self->_set_error( undef );
    
    if ( $res->{success} ) {
        return $res->{content};
    }

    $self->_set_error( $res->{reason} );

    return sprintf '<%s_packages></%s_packages>', ( $self->product ) x 2;
}

sub _build_tree {
    my $self = shift;

    $self->_set_error( undef );

    my $tree;
    eval {
        my $parser = XML::LibXML->new->parse_string( $self->content );
        $tree      = $parser->getDocumentElement;
    } or $self->_set_error( $@ );

    return $tree;
}

1;


lib/OPM/Repository.pm
package OPM::Repository;

# ABSTRACT: parse OPM repositories' framework.xml files to search for add ons

use strict;
use warnings;

use Moo;
use List::Util qw(all);
use Scalar::Util qw(blessed);
use Regexp::Common qw(URI);

use OPM::Repository::Source;

our $VERSION = 0.09;

our $ALLOWED_SCHEME = [ 'HTTP', 'file' ];

has sources => ( is => 'ro', required => 1, isa => sub {
    die "no valid URIs" unless 
        ref $_[0] eq 'ARRAY' 
        and
        all { _check_uri( $_ ) } @{ $_[0] }
});

has _objects => ( is => 'ro', isa => sub {
    die "no valid objects" unless 
        ref $_[0] eq 'ARRAY' 
        and
        all { blessed $_ and $_->isa( 'OPM::Repository::Source' ) } @{ $_[0] }
});

sub find {
    my ($self, %params) = @_;

    my @found;
    for my $source ( @{ $self->_objects || [] } ) {
        my $found = $source->find( %params );
        push @found, $found if $found;
    }

    return @found;
}

sub list {
    my ($self, %params) = @_;

    my %found_packages;
    my @detailed_list;
    for my $source ( @{ $self->_objects || [] } ) {
        my @found = $source->list( %params );
        @found_packages{@found} = (1) x @found;
        push @detailed_list, @found;
    }

    if ( $params{details} ) {
        return sort { $a->{name} cmp $b->{name} || $a->{version} cmp $b->{version} }@detailed_list;
    }

    return sort keys %found_packages;
}

sub BUILDARGS {
    my ($class, @args) = @_;

    unshift @args, 'sources' if @args % 2;

    my %param = @args;

    for my $url ( @{ $param{sources} || [] } ) {
        push @{ $param{_objects} }, OPM::Repository::Source->new( url => $url );
    }

    return \%param;
}

sub _check_uri {
    my @allowed_schemes = ref $ALLOWED_SCHEME ? @{ $ALLOWED_SCHEME } : $ALLOWED_SCHEME;

    my $matches;

    SCHEME:
    for my $scheme ( @allowed_schemes ) {
        my $regex = ( lc $scheme eq 'http' ) ?
            $RE{URI}{HTTP}{-scheme => qr/https?/} :
            $RE{URI}{$scheme};

        if ( $_[0] =~ m{\A$regex\z} ) {
            $matches++;
            last SCHEME;
        }
    }

    die "No valid URI" unless $matches;
    return 1;
}

1;

=head1 SYNOPSIS

  use OPM::Repository;
  
  my $repo = OPM::Repository->new(
      sources => [qw!
          https://opar.perl-services.de/framework.xml
          https://download.znuny.org/releases/packages/framework.xml
          https://download.znuny.org/releases/itsm/packages6/framework.xml
      !],
  );
  
  my ($url) = $repo->find(
    name      => 'ITSMCore',
    framework => '3.3',
  );
  
  print $url;

=head1 METHODS

=head2 new

C<new> has only one mandatory parameter: I<sources>. This has to be 
an array reference of URLs for repositories' framework.xml files.

  my $repo = OPM::Repository->new(
      sources => [qw!
          http://opar.perl-services.de/framework.xml
          http://ftp.framework.org/pub/framework/packages/framework.xml
          http://ftp.framework.org/pub/framework/itsm/packages33/framework.xml
      !],
  );

=head2 find

Search for an add on for a given OPM version in those repositories. It
returns a list of urls if the add on was found, C<undef> otherwise.

  my @urls = $repo->find(
    name      => 'ITSMCore',
    framework => '3.3',
  );

Find a specific version

  my @urls = $repo->find(
    name      => 'ITSMCore',
    framework => '3.3',
    version   => '1.4.8',
  );

=head2 list

List all addons found in the repositories

  my @addons = $repo->list;
  say $_ for @addons;

You can also define the OPM version

  my @addons = $repo->list( framework => '5.0.x' );
  say $_ for @addons;

Both snippets print a simple list of addon names. If you want to
to create a list with more information, you can use

  my @addons = $repo->list(
      framework => '5.0.x',
      details   => 1,
  );
  say sprintf "%s (%s) on %s\n", $_->{name}, $_->{version}, $_->{url} for @addons;
